File: Sources/AvivCore/MarkdownMinimapViewport.swift
import AppKit

public struct MarkdownMinimapViewportMetrics: Equatable {
    public let trackRect: NSRect
    public let documentHeight: CGFloat
    public let visibleMinY: CGFloat
    public let visibleHeight: CGFloat
    public let thumbRect: NSRect

    public var visibleMaxY: CGFloat {
        visibleMinY + visibleHeight
    }

    public var scrollableDocumentHeight: CGFloat {
        max(0, documentHeight - visibleHeight)
    }

    public func projectedY(forDocumentY documentY: CGFloat) -> CGFloat {
        let clampedY = min(max(0, documentY), documentHeight)
        return trackRect.minY + (clampedY / max(1, documentHeight)) * trackRect.height
    }

    public func projectedHeight(forDocumentHeight height: CGFloat) -> CGFloat {
        max(0, height / max(1, documentHeight) * trackRect.height)
    }

    public func documentOffset(forThumbMinY thumbMinY: CGFloat) -> CGFloat {
        guard scrollableDocumentHeight > 0 else { return 0 }
        let scrollableTrackHeight = max(1, trackRect.height - thumbRect.height)
        let clampedY = min(max(trackRect.minY, thumbMinY), trackRect.maxY - thumbRect.height)
        let ratio = (clampedY - trackRect.minY) / scrollableTrackHeight
        return ratio * scrollableDocumentHeight
    }

    public func documentOffset(centeredAtTrackY trackY: CGFloat) -> CGFloat {
        guard scrollableDocumentHeight > 0 else { return 0 }
        let clampedTrackY = min(max(trackRect.minY, trackY), trackRect.maxY)
        let documentCenterY = ((clampedTrackY - trackRect.minY) / max(1, trackRect.height)) * documentHeight
        return min(max(0, documentCenterY - visibleHeight / 2), scrollableDocumentHeight)
    }
}

public enum MarkdownMinimapViewport {
    public static func metrics(
        trackBounds: NSRect,
        documentHeight: CGFloat,
        visibleRect: NSRect,
        horizontalInset: CGFloat = 0,
        minimumThumbHeight: CGFloat = 1.5
    ) -> MarkdownMinimapViewportMetrics {
        let trackRect = NSRect(
            x: trackBounds.minX + horizontalInset,
            y: trackBounds.minY,
            width: max(1, trackBounds.width - horizontalInset * 2),
            height: max(1, trackBounds.height)
        )
        let resolvedDocumentHeight = max(1, documentHeight)
        let visibleHeight = min(resolvedDocumentHeight, max(0, visibleRect.height))
        let maxVisibleMinY = max(0, resolvedDocumentHeight - visibleHeight)
        let visibleMinY = min(max(0, visibleRect.minY), maxVisibleMinY)
        let visibleMaxY = visibleMinY + visibleHeight

        let projectedMinY = trackRect.minY + (visibleMinY / resolvedDocumentHeight) * trackRect.height
        let projectedMaxY = trackRect.minY + (visibleMaxY / resolvedDocumentHeight) * trackRect.height
        let exactHeight = max(0, projectedMaxY - projectedMinY)
        let resolvedMinimumHeight = min(trackRect.height, max(0, minimumThumbHeight))

        let thumbHeight: CGFloat
        let thumbY: CGFloat
        if exactHeight >= resolvedMinimumHeight {
            thumbHeight = exactHeight
            thumbY = projectedMinY
        } else {
            thumbHeight = resolvedMinimumHeight
            let projectedCenterY = (projectedMinY + projectedMaxY) / 2
            thumbY = min(
                max(trackRect.minY, projectedCenterY - thumbHeight / 2),
                trackRect.maxY - thumbHeight
            )
        }

        return MarkdownMinimapViewportMetrics(
            trackRect: trackRect,
            documentHeight: resolvedDocumentHeight,
            visibleMinY: visibleMinY,
            visibleHeight: visibleHeight,
            thumbRect: NSRect(
                x: trackRect.minX,
                y: thumbY,
                width: trackRect.width,
                height: thumbHeight
            )
        )
    }
}

File: Windows/src/Aviv.Windows.Core/MarkdownMinimapViewport.cs
namespace Aviv.Windows.Core;

public readonly record struct RectD(double X, double Y, double Width, double Height)
{
    public double MinX => X;
    public double MinY => Y;
    public double MaxX => X + Width;
    public double MaxY => Y + Height;
    public double MidY => Y + Height / 2;
}

public sealed record MarkdownMinimapViewportMetrics(
    RectD TrackRect,
    double DocumentHeight,
    double VisibleMinY,
    double VisibleHeight,
    RectD ThumbRect)
{
    public double VisibleMaxY => VisibleMinY + VisibleHeight;
    public double ScrollableDocumentHeight => Math.Max(0, DocumentHeight - VisibleHeight);

    public double ProjectedY(double documentY)
    {
        var clampedY = Math.Min(Math.Max(0, documentY), DocumentHeight);
        return TrackRect.MinY + clampedY / Math.Max(1, DocumentHeight) * TrackRect.Height;
    }

    public double ProjectedHeight(double documentHeight)
    {
        return Math.Max(0, documentHeight / Math.Max(1, DocumentHeight) * TrackRect.Height);
    }

    public double DocumentOffsetForThumbMinY(double thumbMinY)
    {
        if (ScrollableDocumentHeight <= 0)
        {
            return 0;
        }

        var scrollableTrackHeight = Math.Max(1, TrackRect.Height - ThumbRect.Height);
        var clampedY = Math.Min(Math.Max(TrackRect.MinY, thumbMinY), TrackRect.MaxY - ThumbRect.Height);
        var ratio = (clampedY - TrackRect.MinY) / scrollableTrackHeight;
        return ratio * ScrollableDocumentHeight;
    }

    public double DocumentOffsetCenteredAtTrackY(double trackY)
    {
        if (ScrollableDocumentHeight <= 0)
        {
            return 0;
        }

        var clampedTrackY = Math.Min(Math.Max(TrackRect.MinY, trackY), TrackRect.MaxY);
        var documentCenterY = (clampedTrackY - TrackRect.MinY) / Math.Max(1, TrackRect.Height) * DocumentHeight;
        return Math.Min(Math.Max(0, documentCenterY - VisibleHeight / 2), ScrollableDocumentHeight);
    }
}

public static class MarkdownMinimapViewport
{
    public static MarkdownMinimapViewportMetrics Metrics(
        RectD trackBounds,
        double documentHeight,
        RectD visibleRect,
        double horizontalInset = 0,
        double minimumThumbHeight = 1.5)
    {
        var trackRect = new RectD(
            trackBounds.MinX + horizontalInset,
            trackBounds.MinY,
            Math.Max(1, trackBounds.Width - horizontalInset * 2),
            Math.Max(1, trackBounds.Height));
        var resolvedDocumentHeight = Math.Max(1, documentHeight);
        var visibleHeight = Math.Min(resolvedDocumentHeight, Math.Max(0, visibleRect.Height));
        var maxVisibleMinY = Math.Max(0, resolvedDocumentHeight - visibleHeight);
        var visibleMinY = Math.Min(Math.Max(0, visibleRect.MinY), maxVisibleMinY);
        var visibleMaxY = visibleMinY + visibleHeight;

        var projectedMinY = trackRect.MinY + visibleMinY / resolvedDocumentHeight * trackRect.Height;
        var projectedMaxY = trackRect.MinY + visibleMaxY / resolvedDocumentHeight * trackRect.Height;
        var exactHeight = Math.Max(0, projectedMaxY - projectedMinY);
        var resolvedMinimumHeight = Math.Min(trackRect.Height, Math.Max(0, minimumThumbHeight));

        double thumbHeight;
        double thumbY;
        if (exactHeight >= resolvedMinimumHeight)
        {
            thumbHeight = exactHeight;
            thumbY = projectedMinY;
        }
        else
        {
            thumbHeight = resolvedMinimumHeight;
            var projectedCenterY = (projectedMinY + projectedMaxY) / 2;
            thumbY = Math.Min(Math.Max(trackRect.MinY, projectedCenterY - thumbHeight / 2), trackRect.MaxY - thumbHeight);
        }

        return new MarkdownMinimapViewportMetrics(
            trackRect,
            resolvedDocumentHeight,
            visibleMinY,
            visibleHeight,
            new RectD(trackRect.MinX, thumbY, trackRect.Width, thumbHeight));
    }
}
